python-connectors/kayrros_retrieve-dataset/connector.py

```diff
@@ -0,0 +1,122 @@
+# This file is the actual code for the custom Python dataset kayrros_retrieve-dataset
+
+from dataiku.connector import Connector
+from utils.authentification import get_headers
+import pandas as pd
+import requests
+import logging
+
+logger = logging.getLogger(__name__) 
+
+
+class MyConnector(Connector):
+
+    # A custom Python dataset is a subclass of connector.
+
+    def __init__(self, config, plugin_config):
+
+        Connector.__init__(self, config, plugin_config) 
+
+        # Retrieve credentials and token
+
+        self.username = config.get("username", "")
+        if not self.username:
+            raise ValueError("A Kayrros account is necessary to fetch the data. Please provide a username.")
+
+        self.password = config.get("password", "")
+        if not self.password:
+            raise ValueError("A Kayrros account is necessary to fetch the data. Please provide a password.")
+
+        self.headers = get_headers(self.username, self.password)
+
+        # Retrieve ids of collection and dataset
+
+        self.collection_id = config.get("collection_id", "")
+        if not self.collection_id:
+            raise ValueError("Choosing a collection is necessary to fetch the data. Please provide one in the collection field.")
+
+        self.dataset_id = config.get("dataset_id", "")
+        if not self.password:
+            raise ValueError("Choosing a dataset is necessary to fetch the data. Please provide one in the collection field.")
+
+    def get_read_schema(self):
+
+        # We don't specify a schema here, so DSS will infer the schema
+        # from the columns actually returned by the generate_rows method
+
+        return None
+
+    def generate_rows(self, dataset_schema=None, dataset_partitioning=None, partition_id=None, records_limit=-1):
+
+        url_asset = "https://platform.api.kayrros.com/v1/processingresult/dataset/" + self.dataset_id
+
+        try:        
+            response = requests.post(url_asset, headers=self.headers)
+            response.raise_for_status()
+
+        except requests.exceptions.RequestException as error:
+            logger.exception("Dataset could not be retrieved because of the following error:\n {}".format(error))
+
+        content = response.json()
+
+        # If no asset, abort it
+
+        nb_assets = len(content["assets"])
+
+        if(nb_assets == 0):
+            raise ValueError("No asset. Please contact Kayrros for access to data.")
+
+        else:
+            logger.info("Received {} assets".format(nb_assets))
+
+        # Generate dataframe
+
+        df = pd.DataFrame(content.get("assets", []))
+
+        # Basic data preparation on results column
+
+        df = df.explode("results").reset_index()
+        results_col = [{**{el: x[el] for el in set(x.keys())-{'metrics'}}, **x["metrics"][0]} for x in df["results"].tolist()]
```

Comment:
It is difficult to see what is going on here... A first step to clarify would be renaming `x` and `el` (`result` and `column_name` ?).
Same with results_col: if it's a column, call it results_column or results_columns

Reply:
solved: [commit](https://github.com/dataiku/dss-plugin-kayrros/commit/8b03c727f2f95e5aa800a905fb42cafe277d41f9)